### ISSUE TITLE
LSPServer doesn't clear diagnostics for deleted files

### ISSUE THREAD

**Issue by yassere** [romejs/rome#652]
## Steps To Reproduce

Run `./scripts/run-vscode-extension` and wait for the Rome extension to output that it's watching.

Add an error to an existing file so that Rome publishes a diagnostic in the `Problems` tab.

Delete that file.

## Results

The diagnostic remains published for that deleted file in the `Problems` tab.

**Comment by sebmck** [comment]
Can you repro with ‘rome lint --watch’? I think you’ll see the same behaviour.

**Comment by yassere** [comment]
Yes, if I run `lint --watch` then delete a file with diagnostics, the diagnostics will stay logged (although the code frame loses the relevant code sample).

**Comment by yassere** [comment]
Is there any way to get a `FileReference` for a deleted path?

A quick and dirty solution for this issue is that `Linter#runWatch` and `LSPServer#watchProject` each create subscriptions to handle their own deletions, but it would be better if `LintRunner#run` just returned diagnostics-less `LintWatchChanges` for deleted files.

Would it make any sense to create an `AbsoluteFilePathMap<FileReference>` of deleted files somewhere that `LintRunner` could use when it hits a `FileNotFound`?

**Comment by sebmck** [comment]
> but it would be better if `LintRunner#run` just returned diagnostics-less `LintWatchChanges` for deleted files.

This should already be happening. Lines 222-229 of `Linter.ts` is already meant to handle deleted paths. The paths that `watchFilesFromArgs` emits includes deleted files. Is `watchFilesFromArgs` not emitting them?

In `LintRunner#computeChanges` we should then be producing a changes object for all updated paths, and so it should then produce an empty diagnostics array.

There's already code to handle all this there must just be a bug somewhere.

**Comment by yassere** [comment]
I think those lines in `Linter.ts` only do anything when a file is deleted while the queue is spinning down. But I don't think you can currently `WorkerQueue#pushQueue` a path that doesn’t exist.

If `LintRunner#run` is being executed as part of a `watchFilesFromArgs` callback in response to a `refreshFileEvent` for a deleted file, `queue.pushQueue(path)` is just going to result in a `FileNotFound` when it tries to get a `WorkerContainer`.

I added the `allowMissing` to prevent it from crashing, but I didn’t notice at the time that we were left with uncleared diagnostics.

I'm also not sure if the `updatedPaths` in `computeChanges` includes deleted paths at the point. I haven't looked at it closely, but wouldn't the call to `projectManager.getFileReference(path)` on line 388  just throw an error if that were the case?

**Comment by sebmck** [comment]
Ahhh that makes sense.

> If `LintRunner#run` is being executed as part of a `watchFilesFromArgs` callback in response to a `refreshFileEvent` for a deleted file, `queue.pushQueue(path)` is just going to result in a `FileNotFound` when it tries to get a `WorkerContainer`.

I think what needs to be done to fix this:

- Add `LintRunner#clearDiagnosticsForPath(path: AbsoluteFilePath)` that contains:

```
this.compilerDiagnosticsCache.set(
	path,
	{suppressions: [], diagnostics: []},
);
```

- Update the `queue.addCallback` callback to use it.

- Change the `pushQueue` wrapping to:

```
for (const path of evictedPaths) {
	const found = await FileNotFound.allowMissing(path, async () => {
		await queue.pushQueue(path);
		return true;
	});

	if (found !== true) {
		this.clearDiagnosticsForPath(path);
	}
}
```

I think that should resolve the remaining issues? I haven't tested it.

> I'm also not sure if the `updatedPaths` in `computeChanges` includes deleted paths at the point. I haven't looked at it closely, but wouldn't the call to `projectManager.getFileReference(path)` on line 388 just throw an error if that were the case?

It should. In `runGraph` the returned paths are basically all the changed file paths + dependents. We do this in case since file updates can cause the dependencies of another file to potentially change to fail or pass.

**Comment by yassere** [comment]
I'd tried that, but I don't think that `runGraph` returns deleted paths in `validatedDependencyPaths` because of the `allowFileNotFound: true` in `graph.seed()` and because of:

```javascript
for (const path of evictedPaths) {
  const newNode = graph.maybeGetNode(path);

  if (newNode === undefined) {
    continue;
  }
  validatedDependencyPaths.add(path);
  // ...
}
```

I think the problem is that `LintRunner#computeChanges` needs a `FileReference` to create a `LintWatchChanges` but if anything calls `projectManager.getFileReference(path)` for a missing path, it just throws an error because `ProjectManager` has already cleaned up in response to `deletedFileEvent`.

That's why I was asking about a `deletedPathToFileReference` map or something (maybe on ProjectManager?) that `LintRunner` could use to produce a `LintWatchChanges` object whenever it sees `FileNotFound` for an evicted path. I don't know whether or not that's a clean solution. I haven't tried to implement it.

**Comment by sebmck** [comment]
Ah, you should just be able to put `validatedDependencyPaths.add(path);` before the `if`, right?

> but if anything calls `projectManager.getFileReference(path)` for a missing path, it just throws an error because `ProjectManager` has already cleaned up in response to `deletedFileEvent`.

What specifically does `getFileReference` call that throws an error? It should just work fine on deleted paths, although it still shouldn't be called.

Regardless, we could just do `this.server.memoryFs.exists` in the loop where we do `changes.push` to handle it.

**Comment by yassere** [comment]
> Ah, you should just be able to put `validatedDependencyPaths.add(path);` before the `if`, right?

I think that would cause problems later with the `graph.validate()` which doesn't use `maybeGetNode`. Plus I think you'd want to rename it if you were adding a path to the set without validating it, hah. But `LintRunner#computeChanges` gets passed `evictedPaths`, so can we just assume that any path in `evictedPaths` but not in `updatedPaths` is a deleted file?

> What specifically does `getFileReference` call that throws an error? 

Hmm, I think it was the `assertProjectExisting` in `getFileReference` that was giving me trouble, but I must have been doing something wrong because that doesn't seem like it should be a problem, now that I take a closer look at it.

>It should just work fine on deleted paths, although it still shouldn't be called.
 
If the change object doesn't contain a `filename` and `FileReference`, then I think the old diagnostics can't be removed by `Linter#runWatch` and `LSPServer` respectively. If we shouldn't call `getFileReference` for a deleted path, how do we create the change object in `computeChanges`?

**Comment by sebmck** [comment]
I found it pretty hard to follow so in the process of investigating I fixed it and opened #672. It's weird because we need to reconcile two different sources of diagnostics. The dependency graph which when validated will return possible cached diagnostics. And then separately diagnostics from the compiler. When we source diagnostics from only the dependency graph and never the compiler then we need to readd them. I'll leave some comments on the linked PR to clarify some things.
